qi-lib/flow/core/compiler/deforest/cps.rkt

```diff
@@ -0,0 +1,371 @@
+             (contract p.contract
+                       (p.prepare
+                        (lambda (state)
+                          (define cstate (inline-consing state rt ...))
+                          cstate)
+                        (#,@#'c.end
+                         (inline-compose1 [t.next t.f
+                                                  '#,(prettify-flow-syntax ctx)
+                                                  '#,(build-source-location-vector
+                                                      (syntax-srcloc ctx))
+                                                  ] ...
+                                          p.next
+                                          )
+                         '#,(prettify-flow-syntax ctx)
+                         '#,(build-source-location-vector
+                             (syntax-srcloc ctx))))
+                       p.name
+                       '#,(prettify-flow-syntax ctx)
+                       #f
+                       '#,(build-source-location-vector
+                           (syntax-srcloc ctx))))))]))
+
+  )
+
+;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
+;; Runtime
+
+(begin-encourage-inline
+
+  ;; Producers
+
+  (define-inline (list->cstream-next done skip yield)
+    (λ (state)
+      (cond [(null? state) (done)]
+            [else (yield (car state) (cdr state))])))
+
+  (define-inline (list->cstream-prepare consing next)
+    (case-lambda
+      [(lst) (next (consing lst))]
+      [rest (void)]))
+
+  (define-inline (range->cstream-next done skip yield)
+    (λ (state)
+      (match-define (list l h s) state)
+      (cond [(< l h)
+             (yield l (cons (+ l s) (cdr state)))]
+            [else (done)])))
+
+  (define-inline (range->cstream-prepare consing next)
+    (case-lambda
+      [(h) (next (consing (list 0 h 1)))]
+      [(l h) (next (consing (list l h 1)))]
+      [(l h s) (next (consing (list l h s)))]
+      [rest (void)]))
+
+  ;; Transformers
+
+  (define-inline (map-cstream-next f next ctx src)
+    (λ (done skip yield)
+      (next done
+            skip
+            (λ (value state)
+              (yield (f value) state)))))
+
+  (define-inline (filter-cstream-next f next ctx src)
+    (λ (done skip yield)
+      (next done
+            skip
+            (λ (value state)
+              (if (f value)
+                  (yield value state)
+                  (skip state))))))
+
+  (define-inline (filter-map-cstream-next f next ctx src)
+    (λ (done skip yield)
+      (next done
+            skip
+            (λ (value state)
+              (let ([fv (f value)])
+                (if fv
+                    (yield fv state)
+                    (skip state)))))))
+
+  (define-inline (take-cstream-next next ctx src)
+    (λ (done skip yield)
+      (λ (take-state)
+        (define n (car take-state))
+        (define state (cdr take-state))
+        (cond ((zero? n)
+               (done))
+              (else
```

Comment:
nit: Seems like `if` would suffice here instead of `cond`

Reply:
Will do.